rust/geoarrow/src/array/coord/array.rs

```diff
@@ -0,0 +1,355 @@
+use std::sync::Arc;
+
+use arrow::array::AsArray;
+use arrow::datatypes::Float64Type;
+use arrow_array::{Array, ArrayRef, FixedSizeListArray, Float64Array, StructArray};
+use arrow_buffer::{Buffer, NullBuffer, ScalarBuffer};
+use arrow_schema::{DataType, Field};
+
+use crate::algorithm::native::eq::coord_eq;
+use crate::array::{CoordType, InterleavedCoordBufferBuilder, SeparatedCoordBufferBuilder};
+use crate::datatypes::{coord_type_to_data_type, Dimension};
+use crate::error::{GeoArrowError, Result};
+use crate::scalar::Coord;
+use crate::trait_::IntoArrow;
+
+/// A GeoArrow coordinate buffer
+#[derive(Clone, Debug)]
+pub struct CoordBuffer {
+    /// We always store 4 buffers in an array, but not all 4 of these may have valid coordinates.
+    /// The number of valid buffers is stored in `num_buffers` and matches the physical size of the
+    /// dimension. For example, `XY` coordinates will only have two valid buffers, in slots `0` and
+    /// `1`. `XYZ` and `XYM` will have three valid buffers and `XYZM` will have four valid buffers.
+    ///
+    /// In the case of interleaved coordinates, each slot will be a clone of the same
+    /// reference-counted buffer.
+    pub(crate) buffers: [ScalarBuffer<f64>; 4],
```

Comment:
I haven't scrutinized the entire PR, but I'd expect any big perf wins from using generics to be here - doing something like `[ScalarBuffer<f64>; T]`

Reply:
For more background, this was a port of @paleolimbot's C implementation [here](https://github.com/geoarrow/geoarrow-c/blob/d353e4d9fd401c298c3cdd67cde5de39922e86c0/src/geoarrow/geoarrow_type.h#L342-L365) and [here](https://github.com/geoarrow/geoarrow-c/blob/d353e4d9fd401c298c3cdd67cde5de39922e86c0/src/geoarrow/geoarrow_type.h#L389-L393).

In GeoArrow we have two ways to store coordinate sequences. Either [`interleaved`](https://github.com/geoarrow/geoarrow/blob/main/format.md#coordinate-interleaved) in one big Vec with all dimensions or [`separated`](https://github.com/geoarrow/geoarrow/blob/main/format.md#coordinate-separated) with a separate vec for each dimension.

Currently GeoArrow uses an [enum CoordBuffer](https://docs.rs/geoarrow/0.4.0-beta.1/geoarrow/array/enum.CoordBuffer.html) with Interleaved and Separated buffers as variants. This means that there's an enum dispatch on **every single coordinate access**. In theory, this should be slow, or at least slower than not having an enum dispatch, I think.

@paleolimbot's C implementation above is able to do this without branching in a clever way. It always stores 4 underlying buffers, but not all 4 of these may have valid coordinates. For interleaved coordinates, it'll store a reference to the _same buffer_ in each position.

Note that here [`ScalarBuffer`](https://docs.rs/arrow/latest/arrow/buffer/struct.ScalarBuffer.html) wraps [`Buffer`](https://docs.rs/arrow/latest/arrow/buffer/struct.Buffer.html), and `Buffer` has an `Arc` under the hood. So storing four of the same cloned buffer should be cheap.

And then here's the meat for coordinate access:

https://github.com/geoarrow/geoarrow-rs/blob/9e6a9538aeaef234e389fdf8cc34f52d1346eaee/rust/geoarrow/src/scalar/coord/scalar.rs#L20-L38

It no longer needs any branching, regardless of whether the underlying Arrow storage is interleaved or separated, by using these "cloned references" and a stride.

I assumed this would be faster, but then the `total_bounds` bench was reliably slower on this PR. Which makes me guess that maybe rustc/llvm is able to do effective branch prediction when it repeatedly calls the same enum dispatch within a single array?